src/refactor_nrepl/externs.clj

```diff
@@ -0,0 +1,28 @@
+(ns refactor-nrepl.externs
+  (:import java.net.URL
+           org.projectodd.shimdandy.ClojureRuntimeShim
+           java.net.URLClassLoader
+           java.util.ArrayList))
+
+(def rt (atom nil))
+
+(defn create-runtime []
+  (let [urls [(URL. "file:libs/shimdandy-impl-1.1.0.jar")
+              (URL. "file:libs/clojure-1.6.0.jar")
+              (URL. "file:libs/pomegranate-0.3.1-SNAPSHOT.jar")]
+        cl (URLClassLoader. (into-array URL urls))]
+    (reset! rt (ClojureRuntimeShim/newRuntime cl "refactor-nrepl"))))
+
+(defn add-dependencies [dep repos]
+  (when-not @rt
+    (create-runtime))
+  (.require @rt (into-array String ["cemerick.pomegranate"]))
```

Comment:
The classloader cannot find pomegranate is because there is no pomegranate jar in lib/ in the current directory. If you are putting the pomegranate jar inside the refactor jar, a URLClassLoader cannot see it when given a file: url, since it's not a file. Unfortunately, you can't give the URLClassLoader a jar: url pointing at the jar within the jar, since it does not support that. If you want to ship jars within jars and then add the inner jars to the classpath, you have to extract them to disk first, then use file: urls that point to the files on disk.

You are correct, Shimdandy can still see the clojure runtime because the parent classloader can see it. Shimdandy doesn't change the way classloader delegation works, so if clojure.jar is visible to a higher classloader, that version gets used. You can't properly use Shimdandy from within a clojure application without writing a custom classloader - it's intended to be used from a java application that then manages multiple runtimes.

For what you are trying to do, it may be easier to use https://github.com/flatland/classlojure, or just create a URLClassLoader, point it to a pomegranate jar somewhere on disk, and call `(.setContextClassLoader (Thread/currentThread) your-cl)` before calling anything that requires pomegranate. You'll probably want to set the TCCL back to what it was originally when you are done.